adr/ADR-19.md

```diff
@@ -0,0 +1,281 @@
+# JetStream based Object Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-11-03|
+|Author  |@scottf|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, objectstore|
+
+## Context
+
+This document describes a design of a JetStream backed object store.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+- Represent an object store.
+- Store a large quantity of related bytes in chunks as a single object.
+- Retrieve all the bytes from a single object
+- Store meta data regarding each object
+- Store multiple objects in a single store  
+- Ability to specify chunk size
+- Ability to delete an object
+- Ability to understand the state of the object store.
+
+Possible future features
+
+- Event Notifications (add/delete/lock)
+- Locking
+- Archiving (tiered storage)
+- Searching/Indexing (tagging)
+- Versioning / Revisions
+- Overriding digest algorithm 
+- Capturing Content-Type (mime type)
+- Per chunk Content-Encoding (i.e. gzip)
+
+## Naming Specification
+
+### Object Store
+The object store name or bucket name (`os-bucket-name`) will be used to formulate a stream name and is specified as: 
+
+`restricted-term` or 1 or more of `A-Z, a-z, 0-9, dash, underscore`
+
+See [ADR-6](ADR-6.md) for more detail. 
+
+### Objects
+An individual object name is used _sanitized_ to form a portion of a subject. The name is specified as:
+
+`limited-term-w-sp (dot limited-term-w-sp)*`
+
+where 
+
+`limited-term-w-sp` is defined as 1 or more of `A-Z, a-z, 0-9, dash, underscore, fwd-slash, equals, space`. 
+
+For an individual object name, a dot `.` is allowed except for the first or last character.
+
+The object name (`os-object-name`) is _sanitized_ by replacing any dot or space in the name with an underscore so it
+can be used in the Meta Info message subject 
+
+> It is under consideration to expand object name to allow printable, which would require some kind of encoding when creating the `os-object-name`. 
```

Comment:
for real world use we'll need this i think, we cant realistically say you cant put `.foo` in a bucket

Reply:
I was just echoing what was implemented in go, but since we are sanitizing the dot anyway for use in the subject, it doesn't really matter if it's at the beginning or end. I'll change this.

Reply:
We are going to change from sanitize to url encoding, which will cover starting or ending with a dot.

Reply:
Just because its in nats.go does not mean its done :) I'd say we have several open areas for discussion in the engineering group, given that it was designed in a vacuum without input.

Reply:
So my comment reflects one of the outstanding design issues I think needs addressing before this should be considered for client parity